activerecord/lib/active_record/delay_touching/state.rb

```diff
@@ -0,0 +1,71 @@
+module ActiveRecord
+  # = Active Record Delay Touching
+  module DelayTouching
+
+    # Tracking of the touch state. This class has no class-level data, so you can
+    # store per-thread instances in thread-local variables.
+    class State # :nodoc:
```

Comment:
Come to think of it this whole business of keeping track of where we are in the update phase for a set of records feels weird to me.

`State` is of two minds; it manages past (`@already_updated_records`) and future (`@records`).

Perhaps it would be better to have `State` act as a linked list.
You could get rid of `@already_updated_records` (+ all that mutation business) and only refer to `@records`. When this state's records have been updated, you simple move on to the next state in the list :grin:

What do you think?

I remember @tenderlove doing something rather elegant like this in Action Dispatch, but I can't find it.